master/buildbot/locks.py

```diff
@@ -164,14 +170,16 @@ def _tryWakeUp(self):
         num_excl, num_counting = self._claimed_excl, self._claimed_counting
         for i, (w_owner, w_access, d) in enumerate(self.waiting):
             if w_access.mode == 'counting':
-                if num_excl > 0 or num_counting >= self.maxCount:
+                if (num_excl > 0 or num_counting >= self.maxCount) \
+                   and w_access.count > 0:
```

Comment:
I think it makes sense to completely short-circuit whole lock logic when dealing with lock accesses with 0 lock count. Just make isAvailable() return true, and `claim()` and `release()` return at the beginning.

Reply:
I changed it accordingly, let me know if that's ok.